[Pages]: Update Mailchannel plugin doc (#5096)
cloudflare/cloudflare-docs: content/pages/platform/functions/plugins/mailchannels.md

The MailChannels Pages Plugin intercepts all form submissions made which have the `data-static-form-name` attribute set. It then emails these form submissions using the MailChannels API.

> The MailChannels API also allows you add a DomainKeys Identified Mail (DKIM) which is an email authentication standard that helps you to sign email messages from your domain with a digital signature using public-key cryptography.

Reviewer: I think this DKIM should follow at the end of the page, rather than be right up at the top. It should also be clear that this is an optional thing.

Probably also worth linking to our [Cloudflare's learning page on DKIM](https://www.cloudflare.com/en-ca/learning/dns/dns-records/dns-dkim-record/) and the [relevant MailChannels documentation](https://mailchannels.zendesk.com/hc/en-us/articles/7122849237389-Adding-a-DKIM-Signature).

> To add a DKIM signature to a message, add the following fields to the personalization object for the message:

Reviewer: ```suggestion
To add a DKIM signature to a message, add the following fields to the `personalization` object for the message:
```

> **dkim_domain**: This is the domain (d=) field for the DKIM signature. To pass DMARC, this should be aligned with the domain in the From header address.
> 
> **dkim_selector**: This is the selector (s=) field for the DKIM signature. It specifies where to find the associated public key in the DNS - see the DKIM specification for more details.
> 
> **dkim_private_key**: The base-64 encoded private key.

Reviewer: ```suggestion
**`dkim_domain`**: This is the domain (`d=`) field for the DKIM signature. To pass DMARC, this should be aligned with the domain in the From header address.

**`dkim_selector`**: This is the selector (`s=`) field for the DKIM signature. It specifies where to find the associated public key in the DNS - see the DKIM specification for more details.

**`dkim_private_key`**: The base-64 encoded private key.
```

## Generate DKIM credentials 

> You can generate your DKIM credentials using any DKIM generator of your choice. To add your generated credentials to your project on the Cloudflare dashboard, do the following:

Reviewer: ```suggestion
You can generate your DKIM credentials using any DKIM generator of your choice. To add your generated credentials to your zone on the Cloudflare dashboard, do the following:
```

> You can generate your DKIM credentials using any DKIM generator of your choice. To add your generated credentials to your project on the Cloudflare dashboard, do the following:

Reviewer: It would be nice to have the OpenSSL commands here to copy-and-paste.

Author: Yea I found a helpful repo, I will add that in.

1. Select the Zone to add DKIM records to. 
2. In the menu on the left select **DNS** > **Add Record**.
3. In the drop-down menu, select **TXT** as the type of record.  
4. Enter your domain name.

> 5. Enter the Public key into your DNS server record as a text(TXT) entry. The name of your DNS record must follow this convention `[selector key]._domainkey.<Your domain>`. 

Reviewer: I think a screenshot here would be useful. The way you enter this into the Cloudflare DNS table is not super intuitive.

You need to omit the zone from the "name" field in the record (as it is automatically added).

This is a record on my zone (don't use this picture since it's a bit different), but you get the idea. Something like this.

<img width="1008" alt="image" src="https://user-images.githubusercontent.com/8484333/178240653-f9a58641-64c4-4cbd-a9fa-3bed4b16aff1.png">

Author: Done!

> 5. Enter the Public key into your DNS server record as a text(TXT) entry. The name of your DNS record must follow this convention `[selector key]._domainkey.<Your domain>`. 

Reviewer: ```suggestion
5. Enter the DKIM record into your DNS server record as a text (TXT) entry. The name of your DNS record must follow this convention `[selector key]._domainkey.<Your domain>`. 
```

> 6. Add the value of your Public key in the content field.

Reviewer: ```suggestion
6. Add the value of your DKIM record in the content field.
```

> 7. Enter the Private key into your email server.

Reviewer: This doesn't happen. There is no email server to add it to. Instead, you add the private key to your Pages Function so that you can send it to MailChannels.

Author: Removed!

The `method` and `action` attributes of the HTML form do not need to be set. The Plugin will automatically override them to allow it to intercept the submission.

For more information about MailChannels and the options they support, refer to [the documentation](https://mailchannels.zendesk.com/hc/en-us/articles/4565898358413-Sending-Email-from-Cloudflare-Workers-using-MailChannels-Send-API).

## DKIM support for Mailchannels API

> The MailChannels API also allows you add a DomainKeys Identified Mail (DKIM) which is an email authentication standard that helps you to sign email messages from your domain with a digital signature using public-key cryptography. Refer to [Cloudflare DNS DKIM guide](https://www.cloudflare.com/en-ca/learning/dns/dns-records/dns-dkim-record/) to learn more.

Author: ```suggestion
The MailChannels API also allows you add a DomainKeys Identified Mail (DKIM) to your DNS records. DKIM is an email authentication standard that helps you to sign email messages from your domain with a digital signature using public-key cryptography. Refer to [Cloudflare DNS DKIM guide](https://www.cloudflare.com/en-ca/learning/dns/dns-records/dns-dkim-record/) to learn more.
```

> The MailChannels API also allows you add a DomainKeys Identified Mail (DKIM) to your DNS records. DKIM is an email authentication standard that helps you to sign email messages from your domain with a digital signature using public-key cryptography. Refer to [Cloudflare DNS DKIM guide](https://www.cloudflare.com/en-ca/learning/dns/dns-records/dns-dkim-record/) to learn more.

Reviewer 2: ```suggestion
The MailChannels API also allows you add a DomainKeys Identified Mail (DKIM) credential to your DNS records. DKIM is an email authentication standard that helps you to sign email messages from your domain with a digital signature using public-key cryptography. Refer to [Cloudflare DNS DKIM guide](https://www.cloudflare.com/en-ca/learning/dns/dns-records/dns-dkim-record/) to learn more.
```
```suggestion
The MailChannels API also allows you add a DomainKeys Identified Mail (DKIM) to your DNS records. DKIM is an email authentication standard that helps you to sign email messages from your domain with a digital signature using public-key cryptography. Refer to [Cloudflare DNS DKIM guide](https://www.cloudflare.com/en-ca/learning/dns/dns-records/dns-dkim-record/) to learn more.
```

The `method` and `action` attributes of the HTML form do not need to be set. The Plugin will automatically override them to allow it to intercept the submission.

For more information about MailChannels and the options they support, refer to [the documentation](https://mailchannels.zendesk.com/hc/en-us/articles/4565898358413-Sending-Email-from-Cloudflare-Workers-using-MailChannels-Send-API).

## DKIM support for Mailchannel API

The MailChannels API also allows you add a DomainKeys Identified Mail (DKIM) which is an email authentication standard that helps you to sign email messages from your domain with a digital signature using public-key cryptography. Refer to [Cloudflare DNS DKIM guide](https://www.cloudflare.com/en-ca/learning/dns/dns-records/dns-dkim-record/) to learn more.

{{<Aside type= "note" header="Optional content">}}
When using the MailChannel plugin adding DKIM is optional. 
{{</Aside>}}

To add a DKIM signature to a message, add the following fields to the `personalization` object for the message:

> **`dkim_domain`**: This is the domain (`d=`) field for the DKIM signature. To pass DMARC, this should be aligned with the domain in the From header address.

Reviewer: ```suggestion
**`dkim_domain`**: This is the domain you're sending the email from. For example, if you're sending an email from `support@example.com`, set this to `example.com`.
```

> **`dkim_selector`**: This is the selector (`s=`) field for the DKIM signature. It specifies where to find the associated public key in the DNS - see the DKIM specification for more details.

Reviewer: ```suggestion
**`dkim_selector`**: This specifies where to find the associated public key in your DNS records. For example, if you make the DKIM record available at `mailchannels._domainkey.example.com`, set this to `mailchannels`.
```

**`dkim_private_key`**: The base-64 encoded private key.

## Generate DKIM credentials 

You can generate your DKIM credentials using [OpenSSL](https://www.openssl.org/) in the following steps:

> 1. Generate your private key and DNS credentials by running the command below in your terminal: 

Reviewer: ```suggestion
1. Generate your private key and DNS record by running the command below in your terminal: 
```

{{<Aside type= "note" header="Optional content">}}
Setting up DKIM for use in your MailChannels Pages Plugin is optional, but will improve the deliverability of your emails. Clients will be less likely to mark your emails as spam or junk because, when configured correctly, DKIM allows clients to verify that you own the domain you're sending from.
{{</Aside>}}

To add a DKIM signature to a message, add the following fields to the `personalization` object for the message:

**`dkim_domain`**: This is the domain you're sending the email from. For example, if you're sending an email from `support@example.com`, set this to `example.com`.

**`dkim_selector`**: This specifies where to find the associated public key in your DNS records. For example, if you make the DKIM record available at `mailchannels._domainkey.example.com`, set this to `mailchannels`.

**`dkim_private_key`**: The base-64 encoded private key.

## Generate DKIM credentials 

You can generate your DKIM credentials using [OpenSSL](https://www.openssl.org/) in the following steps:

1. Generate your private key and DNS record by running the command below in your terminal: 

```sh

> $ openssl genrsa 2048 | tee private_key.pem | openssl rsa -outform der | openssl base64 -A 

Reviewer: ```suggestion
$ openssl genrsa 2048 | openssl rsa -outform der | openssl base64 -A > private_key.txt
```

```
{{<Aside type="note" header="Command breakdown">}}

> `openssl genrsa 2048` generates a 2048-bit RSA key. The output is passed to
> 
> `tee private_key.pem`, which writes the key to the `private_key.pem` file, and passes the key to
> 
> `openssl rsa -outform der | openssl base 64 -A`, which converts the key from PEM format to DER format, then base64 encodes it (this essentially removes the header from the PEM formatted key).

Reviewer: ```suggestion
`openssl genrsa 2048` generates a 2048-bit RSA key. The output is passed to

`openssl rsa -outform der | openssl base 64 -A`, which converts the key from PEM format to DER format, then base64 encodes it. Finally, we take this output into

`> private_key.txt` which saves the contents to `private_key.txt`.
```

```
{{<Aside type="note" header="Command breakdown">}}

This creates a public key from the private key (`openssl rsa -in priv_key.pem -pubout -outform der`), encodes it in base64 (`openssl base 64 -A`), and finally writes it to the `dkim_record.txt` file.

{{</Aside>}}

2. Copy the contents of the `private_key.pem` file and add that as an environment variable to your Pages project by logging into the [Cloudflare dashboard](https://dash.cloudflare.com/login) > **Pages** > select project > **Settings** > **Environment Variables** section > click **Add variables** and add variable name as `DKIM_PRIVATE_KEY` and the content of `private_key.pem` as Value.

3. Create a DNS record with the content of the generated `dkim_record.txt` file content.

Next, look in your generated `dkim_record.txt` file for your DKIM credentials, and add them to your Zone on the Cloudflare dashboard. Follow the steps below:

1. Select the Zone to add DKIM records. 
2. In the menu on the left select **DNS** > **Add Record**.
3. In the drop-down menu, select **TXT** as the type of record.  

> 4. Enter your domain name in the name field.

Reviewer: ```suggestion
```

5. Enter the DKIM record into your DNS server as a text (`TXT`) entry. The name of your DNS record must follow this convention `<selector key>._domainkey`. For example, `mailchannels._domainkey`.

{{<Aside type= "note" header="Selector value">}}
You can choose any value as the selector, as long as it is permitted as a DNS hostname (that is, all lowercase letters, numbers and hyphens).
{{</Aside>}}

> 6. Add the content of your `dkim_record.txt` file in the content field.

Reviewer: ```suggestion
5. Add the content of your `dkim_record.txt` file in the content field.
```

![Follow the instructions above to add DKIM credentials to your Zone DNS records](/pages/platform/functions/plugins/mailchannel_DKIM_DNS_setup.png)

## Add DKIM fields to personalization object

After generating DKIM records, you must add the corresponding fields to the `personalizations` object to use DKIM. 

> The required fields are `dkim_domain`,`dkim_selector`, `dkim_private_key`. The value of these fields must match the values you generated earlier.

Reviewer: ```suggestion
The required fields are `dkim_domain`, `dkim_selector`, `dkim_private_key`. The value of these fields must match the values you generated earlier.
```

The following code block shows an example of using DKIM with the MailChannels Pages Plugin.

```typescript
---
filename: functions/_middleware.ts
highlight: [7,8,9]
---
import mailChannelsPlugin from "@cloudflare/pages-plugin-mailchannels";

> export const onRequest: PagesFunction = mailChannelsPlugin({
>   personalizations: [
>     {
>       to: [{ name: "Some User", email: "user@cloudflare.com" }],
>       "dkim_domain": "example.com", // The value has to be the domain you added DKIM records to and where you're sending your email from
>       "dkim_selector": "mailchannels",
>       "dkim_private_key": env.DKIM_PRIVATE_KEY
>     },
>   ],
>   from: {
>     name: "ACME Support",
>     email: "support@example.com",
>   },
>   respondWith: () => {
>     return new Response(
>       `Thank you for submitting your enquiry. A member of the team will be in touch shortly.`
>     );
>   },
> });

Reviewer: ```suggestion
export const onRequest: PagesFunction = (context) => mailChannelsPlugin({
  personalizations: [
    {
      to: [{ name: "Some User", email: "user@cloudflare.com" }],
      "dkim_domain": "example.com", // The value has to be the domain you added DKIM records to and where you're sending your email from
      "dkim_selector": "mailchannels",
      "dkim_private_key": context.env.DKIM_PRIVATE_KEY
    },
  ],
  from: {
    name: "ACME Support",
    email: "support@example.com",
  },
  respondWith: () => {
    return new Response(
      `Thank you for submitting your enquiry. A member of the team will be in touch shortly.`
    );
  },
})(context);
```

{{</Aside>}}

To add a DKIM signature to a message, add the following fields to the `personalization` object for the message:

**`dkim_domain`**: This is the domain you're sending the email from. For example, if you're sending an email from `support@example.com`, set this to `example.com`.

**`dkim_selector`**: This specifies where to find the associated public key in your DNS records. For example, if you make the DKIM record available at `mailchannels._domainkey.example.com`, set this to `mailchannels`.

**`dkim_private_key`**: The base-64 encoded private key.

## Generate DKIM credentials 

You can generate your DKIM credentials using [OpenSSL](https://www.openssl.org/) in the following steps:

1. Generate your private key and DNS record by running the command below in your terminal: 

```sh
$ openssl genrsa 2048 | tee private_key.pem | openssl rsa -outform der | openssl base64 -A > private_key.txt
```
{{<Aside type="note" header="Command breakdown">}}

> `openssl genrsa 2048` generates a 2048-bit RSA key. The output is passed to
> 
> `tee private_key.pem`, which writes the key to the `private_key.pem` file, and passes the key to
> 
> `openssl rsa -outform der | openssl base 64 -A`, which converts the key from PEM format to DER format, then base64 encodes it. Finally, we take this output into
> 
> `> private_key.txt` which saves the contents to `private_key.txt`.

Reviewer 2: ```suggestion
`openssl genrsa 2048` generates a 2048-bit RSA key. The output is passed to `tee private_key.pem`, which writes the key to the `private_key.pem` file, and passes the key to `openssl rsa -outform der | openssl base 64 -A`, which converts the key from PEM format to DER format, then base64 encodes it. Take this output into `> private_key.txt` which saves the contents to `private_key.txt`.
```

{{</Aside>}}

```sh
$ echo -n "v=DKIM1;p=" > dkim_record.txt && openssl rsa -in private_key.pem -pubout -outform der | openssl base64 -A >> dkim_record.txt
```

> {{<Aside type="note" header="Command breakdown">}}

Reviewer 2: ```suggestion
```

This creates a public key from the private key (`openssl rsa -in priv_key.pem -pubout -outform der`), encodes it in base64 (`openssl base 64 -A`), and finally writes it to the `dkim_record.txt` file.

> {{</Aside>}}
> 

Reviewer 2: ```suggestion
```

> 2. Copy the contents of the `private_key.txt` file and add that as an environment variable to your Pages project by logging into the [Cloudflare dashboard](https://dash.cloudflare.com/login) > **Pages** > select project > **Settings** > **Environment Variables** section > click **Add variables**. Set the variable name as `DKIM_PRIVATE_KEY` and the value as the contents of `private_key.txt` file.

Reviewer 2: ```suggestion
2. Copy the contents of the `private_key.txt` file and add that as an environment variable to your Pages project by logging into the [Cloudflare dashboard](https://dash.cloudflare.com/login) > **Pages** > your Pages project > **Settings** > **Environment Variables* > **Add variables**. Set the variable name as `DKIM_PRIVATE_KEY` and the value as the contents of `private_key.txt` file.
```

3. Create a DNS record with the content of the generated `dkim_record.txt` file content.

> 

Reviewer 2: ```suggestion
```

> Next, look in your generated `dkim_record.txt` file for your DKIM credentials, and add them to your Zone on the Cloudflare dashboard. Follow the steps below:

Reviewer 2: ```suggestion
Next, look in your generated `dkim_record.txt` file for your DKIM credentials, and add them to your website in the Cloudflare dashboard. Follow the steps below:
```

> 1. Select the Zone to add DKIM records. 

Reviewer 2: ```suggestion
1. In Account Home, select the website you would like to add a DKIM record. 
```

2. In the menu on the left select **DNS** > **Add Record**.

> 3. In the drop-down menu, select **TXT** as the type of record.  

Reviewer 2: ```suggestion
3. In the dropdown menu, select **TXT** as the type of record.  
```

4. Enter the DKIM record into your DNS server as a text (`TXT`) entry. The name of your DNS record must follow this convention `<selector key>._domainkey`. For example, `mailchannels._domainkey`.

{{<Aside type= "note" header="Selector value">}}
You can choose any value as the selector, as long as it is permitted as a DNS hostname (that is, all lowercase letters, numbers and hyphens).
{{</Aside>}}

5. Add the content of your `dkim_record.txt` file in the content field.

> ![Follow the instructions above to add DKIM credentials to your Zone DNS records](/pages/platform/functions/plugins/mailchannel_DKIM_DNS_setup.png)

Reviewer 2: ```suggestion
![Follow the instructions above to add DKIM credentials to your DNS records](/pages/platform/functions/plugins/mailchannel_DKIM_DNS_setup.png)
```

Look in your generated `dkim_record.txt` file for your generated credentials, and add them to your Zone on the Cloudflare dashboard. Follow the steps below:

1. Select the Zone to add DKIM records. 
2. In the menu on the left select **DNS** > **Add Record**.
3. In the drop-down menu, select **TXT** as the type of record.  
4. Enter your domain name.
5. Enter the DKIM record into your DNS server as a text (TXT) entry. The name of your DNS record must follow this convention `[selector key]._domainkey.<Your domain>`. 

{{<Aside type= "note" header="Value of selector">}}
You can choose any value as the selector, as long as it is permitted to use as a DNS hostname (that is: all lowercase letters, numbers and hyphens). It is recommended to use the name of the service and some date indicator, so it's easy to remember where this key is used. E.g: cloudflare2022.
{{</Aside>}}

6. Add the content of your `dkim_record.txt` file in the content field.
7. Add the private key to your Pages Function.

![Follow the instructions above to add DKIM credentials to your Zone DNS records](/pages/platform/functions/plugins/mailchannel_DKIM_DNS_setup.png)

## Add DKIM fields to personalization object

> 

Reviewer 2: Describe what the user needs to do with this code block.
